### PR TITLE
https support for proxy relay

### PR DESCRIPTION
**Pull request by arenevier**
Fix for issue #563

### CODE REVIEW

**Review by jancurn — APPROVED** [review]
Thanks for the improvement, it looks like a reasonable change. Any chance you could also update the tests?

**Comment by arenevier** [comment]
Would be happy to modify the tests. But I'm not sure what to modify. Also, running the tests hang on my laptop, after the following output:

```

% npm test

> proxy-chain@2.5.5 test
> nyc cross-env NODE_OPTIONS=--insecure-http-parser mocha --bail



(node:232095) Warning: Using insecure HTTP parsing
(Use `node --trace-warnings ...` to show where the warning was created)
  ✔ supports localAddress
  ✔ supports custom CONNECT server handler
  ✔ supports pre-response CONNECT payload
  utils.anonymizeProxy
    ✔ throws for invalid args
    ✔ throws for unsupported https: protocol
    ✔ throws for invalid ports
    ✔ throws for invalid URLs
    ✔ keeps already anonymous proxies (both with callbacks and promises)
    1) anonymizes authenticated upstream proxy (both with callbacks and promises)


  8 passing (63ms)
  1 failing

  1) utils.anonymizeProxy
       anonymizes authenticated upstream proxy (both with callbacks and promises):
     AssertionError: expected 'socket hang up' to include 'ECONNREFUSED'
      at /home/arno/src/proxy-chain/test/anonymize_proxy.js:236:48
      at processTicksAndRejections (node:internal/process/task_queues:95:5)

```

**Comment by arenevier** [comment]
I added a test. I was not able to connect to a https server in the test. But in the test, I can check that proxyServer does not consider the relay url as invalid. The code snippet will fail without the PR, and pass with it.

File: test/server.js [hunk 1]
```diff
@@ -1337,6 +1338,45 @@ it('supports localAddress', async () => {
     }
 });
 
+it('supports https proxy relay', async () => {
+    const target = https.createServer((_req, _res) => {
```
Comment (jirimoravcik) [inline test/server.js:1342]:
```suggestion
    const target = https.createServer(() => {
```

File: test/server.js [hunk 2]
```diff
@@ -1337,6 +1338,45 @@ it('supports localAddress', async () => {
...
+        port: 6666,
+        prepareRequestFunction: () => {
+            console.log(`https://localhost:${target.address().port}`);
+            return {
+                upstreamProxyUrl: `https://localhost:${target.address().port}`,
+            }
+        }
+    })
+    let proxyServerError = false;
+    proxy.on('requestFailed', (_err) => {
```
Comment (jirimoravcik) [inline test/server.js:1357]:
```suggestion
    proxy.on('requestFailed', () => {
```

File: test/server.js [hunk 3]
```diff
@@ -1337,6 +1338,45 @@ it('supports localAddress', async () => {
...
+    target.listen(() => {
+    });
+
+    const proxy = new ProxyChain.Server({
+        port: 6666,
+        prepareRequestFunction: () => {
+            console.log(`https://localhost:${target.address().port}`);
+            return {
+                upstreamProxyUrl: `https://localhost:${target.address().port}`,
+            }
```
Comment (jirimoravcik) [inline test/server.js:1353]:
```suggestion
            };
```

File: test/server.js [hunk 4]
```diff
@@ -1337,6 +1338,45 @@ it('supports localAddress', async () => {
...
+        }
```
Comment (jirimoravcik) [inline test/server.js:1354]:
```suggestion
        },
```

File: test/server.js [hunk 5]
```diff
@@ -1337,6 +1338,45 @@ it('supports localAddress', async () => {
...
+    })
```
Comment (jirimoravcik) [inline test/server.js:1355]:
```suggestion
    });
```

File: test/server.js [hunk 6]
```diff
@@ -1337,6 +1338,45 @@ it('supports localAddress', async () => {
...
+        });
+    } catch (e) {
+        // the request will fail with the following error:
+        // Error: tunneling socket could not be established, statusCode=599
+    }
+    expect(proxyServerError).to.be.equal(false);
+
+    proxy.close();
+    target.close();
+})
```
Comment (jirimoravcik) [inline test/server.js:1378]:
```suggestion
});
```

File: test/server.js [hunk 7]
```diff
@@ -1337,6 +1338,45 @@ it('supports localAddress', async () => {
     }
 });
 
+it('supports https proxy relay', async () => {
+    const target = https.createServer((_req, _res) => {
+    });
+    target.listen(() => {
+    });
+
+    const proxy = new ProxyChain.Server({
```
Comment (jirimoravcik) [inline test/server.js:1347]: please rename this to `proxyServer`

**Review by jirimoravcik — CHANGES_REQUESTED** [review]

**Review by jirimoravcik — APPROVED** [review]

**Pull request merged**